apps/example-todo-app/pages/api/todo/auth-token-or-user-session.ts

```diff
@@ -0,0 +1,13 @@
+import { withRouteSpec } from "lib/middlewares"
+import { z } from "zod"
+
+export default withRouteSpec({
+  methods: ["GET"],
+  auth: ["auth_token", "user_session"],
```

Comment:
the one downside of this syntax is that (without any additional information) if none of the auth middlewares succeed, the error returned will be from the last middleware in the list

currently in seam connect's `with-api-key-or-session-or-token.ts` we have this logic to avoid throwing an error specific to an auth method that the consumer isn't actually using:

```ts
  const token = (req.headers.authorization?.split("Bearer ")?.[1] ??
    req.headers["seam-client-session-token"] ??
    req.headers["client-session-token"]) as string | null

  if (!token) {
    throw new UnauthorizedException({
      type: "unauthorized",
      message: "No token found in header (did you mean to add Authorization?)",
    })
  }
```

open to ideas on how to make this interface better

-------------------

alternatively, maybe the easiest solution is to suggest consumers add an additional auth "method" like this:

```ts
const withRouteSpec = createWithRouteSpec({
  authMiddlewareMap: {
    has_token_header: (next) => async (req, res) => {
      const token = (req.headers.authorization?.split("Bearer ")?.[1]) as string | null

      if (!token) {
        throw new UnauthorizedException({
          type: "unauthorized",
          message: "No token found in header",
        })
      }

      throw new Error("Throwing to trigger the next auth middleware (has_token_header should never be the only auth middleware)")
    },
    access_token: (next) => async (req, res) => {
      const token = (req.headers.authorization?.split("Bearer ")?.[1]) as string | null

      if (!token) {
        // Throw specific errors
        throw new UnauthorizedException({
          type: "access_token_unauthorized",
          message: "No access token found in header",
        })
      } 
    },
    api_key: // ...
  }
})

withRouteSpec({
  auth: ['has_token_header', 'access_token', 'api_key'],
})
```

Reply:
I think we should allow createWithRouteSpec to accept `onMultipleAuthFailure(errors)`, likewise with `withRouteSpec`, not a huge deal in the short term

Reply:
yeah, that sounds good

Reply:
d449ca7